scada/scada.go

```diff
@@ -0,0 +1,238 @@
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	sc "github.com/hashicorp/scada-client"
+)
+
+const (
+	InfrastructureResource = "infrastructures"
+	BoxesResource          = "boxes"
+	SharesResource         = "shares"
+	NomadClusterResource   = "nomad-cluster"
+)
+
+// Provider wraps scada-client.Provider to allow most applications to only pull
+// in this package
+type Provider struct {
+	*sc.Provider
+}
+
+type AtlasConfig struct {
+	// Endpoint is the SCADA endpoint used for Atlas integration. If empty, the
+	// defaults from the provider are used.
+	Endpoint string `mapstructure:"endpoint"`
+
+	// The name of the infrastructure we belong to, e.g. "hashicorp/prod"
+	Infrastructure string `mapstructure:"infrastructure"`
+
+	// The Atlas authentication token
+	Token string `mapstructure:"token" json:"-"`
+}
+
+// Config holds the high-level information used to instantiate a SCADA provider
+// and listener
+type Config struct {
+	// The service name to use
+	Service string
+
+	// The version of the service
+	Version string
+
+	// The type of resource we represent
+	ResourceType string
+
+	// Metadata to send to along with the service information
+	Meta map[string]string
+
+	// If set, TLS certificate verification will be skipped. The value of the
+	// SCADA_INSECURE environment variable will be considered if this is false.
+	// If using SCADA_INSECURE, any non-empty value will trigger insecure mode.
+	Insecure bool
+
+	// Holds Atlas configuration
+	Atlas AtlasConfig
+}
+
+// ProviderService returns the service information for the provider
+func providerService(c *Config) *sc.ProviderService {
+	ret := &sc.ProviderService{
+		Service:        c.Service,
+		ServiceVersion: c.Version,
+		Capabilities:   map[string]int{},
+		Meta:           c.Meta,
+		ResourceType:   c.ResourceType,
+	}
+
+	return ret
+}
+
+// providerConfig returns the configuration for the SCADA provider
+func providerConfig(c *Config) *sc.ProviderConfig {
+	ret := &sc.ProviderConfig{
+		Service:       providerService(c),
+		Handlers:      map[string]sc.CapabilityProvider{},
+		Endpoint:      c.Atlas.Endpoint,
+		ResourceGroup: c.Atlas.Infrastructure,
+		Token:         c.Atlas.Token,
+	}
+
+	return ret
+}
+
+// NewProvider creates a new SCADA provider using the given configuration.
+// Requests for the HTTP capability are passed off to the listener that is
+// returned.
+func NewHTTPProvider(c *Config, logOutput io.Writer) (*Provider, net.Listener, error) {
+	// Get the configuration of the provider
+	config := providerConfig(c)
+	config.LogOutput = logOutput
+
+	// Set the HTTP capability
+	config.Service.Capabilities["http"] = 1
+
+	// SCADA_INSECURE env variable is used for testing to disable TLS
+	// certificate verification.
+	insecure := c.Insecure
+	if !insecure {
+		if os.Getenv("SCADA_INSECURE") != "" {
+			insecure = true
+		}
+	}
+	if insecure {
+		config.TLSConfig = &tls.Config{
+			InsecureSkipVerify: true,
+		}
+	}
```

Comment:
I think we can move the `SCADA_INSECURE` handling into `providerConfig()` since its not HTTP-specific